pkg/cgroup/linux.go

```diff
@@ -0,0 +1,213 @@
+
+		tokens := strings.SplitN(line, ":", 3)
+		if len(tokens) < 3 {
+			continue
+		}
+
+		_, name, path := tokens[0], tokens[1], tokens[2]
+		for _, token := range strings.Split(name, ",") {
+			name = strings.TrimPrefix(token, "name=")
+			proc[name] = []string{path}
+		}
+	}
+
+	// Return upon any error while reading the cgroup categories.
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return proc, nil
+}
+
+// Control group manager CLI defaults.
+const (
+	// Change this if at all in future control group manager
+	// changes their cli name.
+	cgmProgName = "cgm"
+
+	// Action param name used to fetch value for a particular
+	// controller.
+	cgmProgGetValue = "getvalue"
+)
+
+// Fetch value of the cgroup kernel param from the cgroup manager,
+// if cgroup manager is configured we should just rely on `cgm` cli
+// to fetch all the values for us.
+func getCGroupManagerKernValue(cname, path, kernParam string) (limit uint64, err error) {
+	// Default value for an unconfigured cgroup memory limit.
+	//
+	//    $ cat /sys/fs/cgroup/memory/memory.limit_in_bytes
+	//    9223372036854771712
+	//
+	limit = unlimitedMemoryLimit
+
+	cmd := exec.Command(cgmProgName, cgmProgGetValue, cname, path, kernParam)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err = cmd.Run(); err != nil {
+		return limit, err
+	}
+
+	var v uint64
+	v, err = strconv.ParseUint(strings.TrimSpace(out.String()), 10, 64)
+	if err != nil {
+		return limit, err
+	}
+
+	limit = v
+	return limit, nil
+}
+
+// Get cgroup memory limit file path.
+func getCGroupMemoryLimitFilePath(cgPath string) string {
+	path := cgroupMemSysPath
+
+	// Docker generates weird cgroup paths that don't
+	// really exist on the file system.
+	//
+	// For example on regular Linux OS :
+	// `/user.slice/user-1000.slice/session-1.scope`
+	//
+	// But they exist as a bind mount on Docker and
+	// are not accessible :  `/docker/<hash>`
+	//
+	// We we will just ignore if there is `/docker` in the
+	// path ignore and fall back to :
+	// `/sys/fs/cgroup/memory/memory.limit_in_bytes`
+	if !strings.HasPrefix(cgPath, dockerPrefixName) {
+		path = filepath.Join(path, cgPath)
+	}
+
+	// Final path.
+	return filepath.Join(path, memoryLimitKernelParam)
+}
+
+// GetCGroupMemoryLimit - Fetches cgroup memory limit either
+// from a file path at '/sys/fs/cgroup/memory', if path fails then
+// fallback to querying cgroup manager.
+func GetCGroupMemoryLimit(pid int) (limit uint64, err error) {
+	limit = unlimitedMemoryLimit
+
+	var cg CGEntries
+	cg, err = GetProcCGroup(pid)
+	if err != nil {
+		return limit, err
+	}
+
+	path := cg.GetCGPathByName("memory")
+
+	var v uint64
+	v, err = getCGroupManagerKernValue("memory", path, memoryLimitKernelParam)
+	if err != nil {
+
+		var b []byte
+		b, err = ioutil.ReadFile(getCGroupMemoryLimitFilePath(path))
+		if err != nil {
+			return limit, err
+		}
+
+		v, err = strconv.ParseUint(strings.TrimSpace(string(b)), 10, 64)
+		if err != nil {
+			return limit, err
+		}
```

Comment:
Below code is enough, as no one supposes to use `limit` in case of error is set.

```go
if limit, err = getCGroupManagerKernValue("memory", path, memoryLimitKernelParam); err == nil {
    return limit, err
}

var b []byte
if b, err = ioutil.ReadFile(getCGroupMemoryLimitFilePath(path)); err == nil {
    limit, err = strconv.ParseUint(strings.TrimSpace(string(b)), 10, 64)
}
return limit, err
```